File: leetcode/tree/429.n叉树的层序遍历.js
/*
 * @lc app=leetcode.cn id=429 lang=javascript
 *
 * [429] N叉树的层序遍历
 *
 * https://leetcode-cn.com/problems/n-ary-tree-level-order-traversal/description/
 *
 * algorithms
 * Medium (66.46%)
 * Likes:    109
 * Dislikes: 0
 * Total Accepted:    29.7K
 * Total Submissions: 44.7K
 * Testcase Example:  '[1,null,3,2,4,null,5,6]'
 *
 * 给定一个 N 叉树，返回其节点值的层序遍历。 (即从左到右，逐层遍历)。
 *
 * 例如，给定一个 3叉树 :
 *
 *
 *
 *
 *
 *
 *
 * 返回其层序遍历:
 *
 * [
 * ⁠    [1],
 * ⁠    [3,2,4],
 * ⁠    [5,6]
 * ]
 *
 *
 *
 *
 * 说明:
 *
 *
 * 树的深度不会超过 1000。
 * 树的节点总数不会超过 5000。
 *
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number[][]}
 */
// recursive
var levelOrder = function (root) {
  const levels = [];
  dfs(root, 0);
  return levels;

  function dfs(node, level) {
    if (!node) return;
    if (levels.length === level) levels.push([]);
    levels[level].push(node.val);
    if (node.children.length) {
      for (const child of node.children) {
        dfs(child, level + 1);
      }
    }
  }
};

// iterative
var levelOrder = function (root) {
  const levels = [];
  if (!root) return levels;
  const queue = [];
  queue.push(root);
  let level = 0;

  while (queue.length) {
    levels.push([]);
    const size = queue.length;
    for (let i = 0; i < size; i++) {
      const node = queue.shift();
      levels[level].push(node.val);
      if (node.children.length) {
        for (const child of node.children) {
          queue.push(child);
        }
      }
    }
    level++;
  }

  return levels;
};

// iterative 2
var levelOrder = function (root) {
  const levels = [];
  if (!root) return levels;
  const queue = [];
  queue.push(root);

  while (queue.length) {
    const level = [];
    const size = queue.length;
    for (let i = 0; i < size; i++) {
      const node = queue.shift();
      level.push(node.val);
      if (node.children.length) {
        queue.push(...node.children);
      }
    }
    levels.push(level);
  }

  return levels;
};
// @lc code=end

File: leetcode/tree/429.n-叉树的层序遍历.go
/*
 * @lc app=leetcode.cn id=429 lang=golang
 *
 * [429] N 叉树的层序遍历
 *
 * https://leetcode-cn.com/problems/n-ary-tree-level-order-traversal/description/
 *
 * algorithms
 * Medium (71.59%)
 * Likes:    231
 * Dislikes: 0
 * Total Accepted:    87K
 * Total Submissions: 121.2K
 * Testcase Example:  '[1,null,3,2,4,null,5,6]'
 *
 * 给定一个 N 叉树，返回其节点值的层序遍历。（即从左到右，逐层遍历）。
 *
 * 树的序列化输入是用层序遍历，每组子节点都由 null 值分隔（参见示例）。
 *
 *
 *
 * 示例 1：
 *
 *
 *
 *
 * 输入：root = [1,null,3,2,4,null,5,6]
 * 输出：[[1],[3,2,4],[5,6]]
 *
 *
 * 示例 2：
 *
 *
 *
 *
 * 输入：root =
 * [1,null,2,3,4,5,null,null,6,7,null,8,null,9,10,null,null,11,null,12,null,13,null,null,14]
 * 输出：[[1],[2,3,4,5],[6,7,8,9,10],[11,12,13],[14]]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 树的高度不会超过 1000
 * 树的节点总数在 [0, 10^4] 之间
 *
 *
 */

package leetcode

// Definition for a Node.
type Node struct {
	Val      int
	Children []*Node
}

// @lc code=start
func levelOrder2(root *Node) [][]int {
	if root == nil {
		return [][]int{}
	}
	var res [][]int
	var queue []*Node
	queue = append(queue, root)
	for len(queue) > 0 {
		var level []int
		size := len(queue)
		for i := 0; i < size; i++ {
			level = append(level, queue[i].Val)
			for _, v := range queue[i].Children {
				queue = append(queue, v)
			}
		}
		res = append(res, level)
		queue = queue[size:]
	}
	return res
}

func levelOrder(root *Node) (res [][]int) {
	var dfs func(root *Node, level int)
	dfs = func(root *Node, level int) {
		if root == nil {
			return
		}
		if len(res) == level {
			res = append(res, []int{})
		}
		res[level] = append(res[level], root.Val)
		for _, v := range root.Children {
			dfs(v, level+1)
		}
	}
	dfs(root, 0)
	return
}

// @lc code=end
